include/deal.II/matrix_free/assemble_operator.h

```diff
@@ -0,0 +1,278 @@
+                    auto c1 = it->column();
+                    auto c2 = it2->column();
+                    if (c1 > c2)
+                      std::swap(c1, c2);
+                    edges.insert(std::make_pair(c1, c2));
+                  }
+              }
+          }
+
+        // Construct a graph from the edge-set.
+        for (const Edge &e : edges)
+          add_edge(e.first, e.second, graph);
+
+        AssertDimension(pattern.n_cols(), boost::num_vertices(graph));
+
+        max_entries_per_row = pattern.max_entries_per_row();
+      }
+
+      /**
+       * Color the existing graph using sequential vertex coloring.
+       * Different orderings of the vertices can give different results.
+       * By default, we do not reorder them, as the benefit does not seem to be
+       * huge.
+       *
+       * @param <tt>order</tt> defines a reording of the vertices.
+       * The given type needs to satisfy the boost::property_map concept.
+       * Example:
+       *   boost::vector_property_map<vertices_size_type> order;
+       *   boost::smallest_last_vertex_ordering(graph, order);
+       */
+      template <
+        typename Order = boost::typed_identity_property_map<vertices_size_type>>
+      void
+      make_coloring(const Order &&order =
+                      boost::typed_identity_property_map<vertices_size_type>{})
+      {
+        color_vec.resize(boost::num_vertices(graph));
+        coloring =
+          boost::make_iterator_property_map(&color_vec.front(),
+                                            get(boost::vertex_index, graph));
+        num_colors = boost::sequential_vertex_coloring(graph, order, coloring);
+
+        entries_per_color.resize(num_colors);
+        for (vertices_size_type color = 0; color < num_colors; color++)
+          {
+            // Columns of color c.
+            entries_per_color[color].reserve(max_entries_per_row);
+            for (vertices_size_type i = 0; i < color_vec.size(); i++)
+              if (color_vec[i] == color)
+                entries_per_color[color].push_back(i);
+          }
+      }
+    };
+
+    /**
+     * Assemble a given operator into a matrix, using the provided sparsity
+     pattern.
+     *
+     * @param <tt>output</tt> resulting matrix. Needs to provide a function set(row, column, value) that allows to write to each entry of the
+     sparsity pattern.
+
+     * @param <tt>op</tt> the operator to assemble. Needs to provide a function vmult(RangeVector&, const DomainVector&).
+
+     * @param <tt>pattern</tt> the underlying sparsity pattern of <tt>op</tt>.
+     * Note that <tt>pattern</tt> can be a subset of the sparsity pattern from
+     <tt>output</tt>
+
+     * @param <tt>dummy_domain</tt> a vector used to initialize vectors in the domain of <tt>op</tt> by copying the structure of
+     <tt>dummy_domain</tt>.
+     * The type <tt>DomainVector</tt> needs to provide reinit(const
+     DomainVector&, bool=false) to create a zero vector.
```

Comment:
I don't like this dummy approach. Cannot we require that the operator provides also `initialize_dof_vector(vec)`?

Reply:
I guess if you want to support non-symmetrical matrices, you could require `initialize_dof_vector(vec, dof_handler_index)` and then you could provide two indices one for the source vector and one for the destination vector.

Reply:
The problem with the `initialize_dof_vector` approach is, that each Operator type (MatrixFree, LinearOperator, ...) would need to have such a method.

Using the dummy vectors, each used vector class only needs to be able to copy its structure to another instance, which I think is more flexible. That way, it is the users responsibility to provide vectors with correct block/parallel partitioning, such that they fit the Operator class.